### ISSUE TITLE
C.43 might be too strict

### ISSUE THREAD

**Issue by ennmichael** [isocpp/CppCoreGuidelines#1119]
The wording of C.43 may lead newbies into trying too hard to provide default constructors for classes where this doesn't make sense. I know that the rule at one point states "so if [a default constructor] can be *reasonably* defined, it should be." However, how reasonable is reasonable? 

Also, I don't feel like C.43 is argumented enough. Quote: "Many language and library facilities rely on default constructors to initialize their elements, e.g. `T a[10]` and `std::vector<T> v(10)`." To me, these examples don't feel strong enough. For vectors we can instead use `std::vector<T>(10, v)`, where `v` is a "default" value that most suites our particular use case (as is mentioned somewhere in the C.43 as well).

Arrays are a good point for discussion. I feel like in the worst case the user would use a function that initializes individual elements in a loop and returns a `std::array` object.

I don't see why anyone would expect a `Date` class to have a default constructor, for example. (Is it perhaps because it has a default constructor in some other language? :-) ). In fact, maybe

    class Date {
    public:
        static Date epoch()
        { ... }

        Date(...) {} // Non-default constructor, or just use list-initialization
    };

would be a better approach. What if person A expects a `Date` class to be initialized to epoch (like in the current example), while person B expects it to be initialized to the current time (this is how `Date` works in Java and JavaScript, so people from those backgrounds would have different habits and expectations)?

To re-iterate:

- Can we define more precisely how far one should go in their efforts to define a default constructor?
- Can we provide stronger reasons as to why one should try to define a default constructor?
- Should we maybe provide an easier example than a `Date` class? Should we maybe change up the current example a bit?

**Comment by neilmacintosh** [comment]
This issue begins with a claim for which I see no evidence. I am not convinced that C.43 is going to lead programmers new to C++ down some difficult path.

C.43 makes a recommendation and clearly calls out it should only be followed when reasonable to do so. The item contains examples of when a reasonable default does not exist, and also walks through how a reasonable default for a Date class can be defined. I'm not really sure what more it could do to prompt the reader to think about the issue and decide for themselves - which is what they must do at the end of the day.

Some guidelines can be very simple and prescriptive, but many - like this one - are about encouraging the reader to think about a principle or approach and how it can be followed. In this case, the key approach is to think about what types in your program might fit the definition of "value type" (a critical aspect of the guideline that I don't see mentioned in this issue) and how a value type works most naturally.

Value types are expected to be moveable, copyable, equality comparable and generally behave like built-in types. Supporting default construction is part of that makeup as it allows them to be used in the same contexts (such as the example of standard library containers) where built-in types can be used. Consistency leads to simpler programs and predictable outcomes. It seems a powerful enough argument to me.

**Comment by AndrewPardoe** [comment]
Herb, could you please take a crack at this one?

**Comment by hsutter** [comment]
We discussed it on the call today, and I suggested (and was asked to implement) clarifying the item this way:

1. Actually define "value type" as copyable and comparable, which removes the current note.

2. Update the examples to match. They're already pretty reasonable but the `Shape` example (which is also missing `virtual` twice) then needs much less rationale -- it's not copyable, so the advice here doesn't apply to it.

I might see some more improvements as I edit it... Thanks!
